# KLM operators for our calculator

First we divide our interaction into two groups - keyboard and mouse. We use a real physical mouse and not the touchpad. With lots and lots of possibilites to combine mouse and keyboard within the calculation we focus on three typical scenarios

-  Using keyboard only
-  Using mouse only
-  Using keyboard to select digits and mouse to select arithmetic operations

We start our experiment with the mouse in the middle of our screen and both hands on the keyboard. We conducted the experiment by ourself and extracted the raw data from stdout to csv. Let's get going!

In [2]:
import numpy as np
import pandas as pd
import datetime as dt

## Using keyboard only

We sub-divide the keyboard tasks into three reasonable KLM operators:

- (1) Pressing a single key (0,1,2,3,4,5,6,7,8,9,+,-)
- (2) Pressing and holding shift to press a single key (*, /)
- (3) Pressing a single key repeatedly (666, 55555)

Let's get our data ready!

In [108]:
df1 = pd.read_csv('raw_data_6.2/press_single_key.csv')

#some lists as placeholders to pass objects
df1_timestamp_data = df1["timestamp"].tolist()
df1_seconds_data = []
df1_delta_data = []

#get seconds with milliseconds from each timestamp
for i in range(len(data)):
    timestamps = data[i]
    seconds = timestamps[17:]
    seconds = float(seconds)
    df1_seconds_data.append(seconds)

#get timedelta between keystrokes    
for i in range(len(df1_seconds_data)):
    delta = df1_seconds_data[i] - df1_seconds_data[i-1]
    df1_delta_data.append(delta)

    
df1_delta_data.pop(0)

np.mean(df1_delta_data)

0.25342900000000007

In [109]:
df2 = pd.read_csv('raw_data_6.2/press_shift_single_key.csv')

df2_timestamp_data = df2["timestamp"].tolist()
df2_seconds_data = []
df2_delta_data = []

for i in range(len(df2_timestamp_data)):
    timestamps = df2_timestamp_data[i]
    seconds = timestamps[17:]
    seconds = float(seconds)
    df2_seconds_data.append(seconds)
    
for i in range(len(df2_seconds_data)):
    delta = df2_seconds_data[i] - df2_seconds_data[i-1]
    df2_delta_data.append(delta)

df2_delta_data.pop(0)

np.mean(df2_delta_data)

0.545387125

In [99]:
df3 = pd.read_csv('raw_data_6.2/press_single_key_multiple_times.csv')

df3_timestamp_data = df3["timestamp"].tolist()
df3_seconds_data = []
df3_delta_data = []

for i in range(len(df3_timestamp_data)):
    timestamps = df3_timestamp_data[i]
    seconds = timestamps[17:]
    seconds = float(seconds)
    df3_seconds_data.append(seconds)

for i in range(len(df3_seconds_data)):
    delta = df3_seconds_data[i] - df3_seconds_data[i-1]
    df3_delta_data.append(delta)
    
df3_delta_data.pop(0)

np.mean(df3_delta_data)

0.2338534081632653

## Using mouse only

We sub-divide the mouse tasks into:

- (1) clicking one digit after another
- (2) clicking an arithmetic operator with a digit entered




In [5]:
df4 = pd.read_csv('raw_data_6.2/click_single_digits.csv')

df4_timestamp_data = df4["timestamp"].tolist()
df4_seconds_data = []
df4_delta_data = []

for i in range(len(df4_timestamp_data)):
    timestamps = df4_timestamp_data[i]
    seconds = timestamps[17:]
    seconds = float(seconds)
    df4_seconds_data.append(seconds)

for i in range(len(df4_seconds_data)):
    delta = df4_seconds_data[i] - df4_seconds_data[i-1]
    df4_delta_data.append(delta)
    
df4_delta_data.pop(0)

np.mean(df4_delta_data)

0.5449099655172415

In [7]:
df5 = pd.read_csv('raw_data_6.2/click_digit_with_arithmetic_operation.csv')

df5_timestamp_data = df5["timestamp"].tolist()
df5_seconds_data = []
df5_delta_data = []

for i in range(len(df5_timestamp_data)):
    timestamps = df5_timestamp_data[i]
    seconds = timestamps[17:]
    seconds = float(seconds)
    df5_seconds_data.append(seconds)

for i in range(len(df5_seconds_data)):
    delta = df5_seconds_data[i] - df5_seconds_data[i-1]
    df5_delta_data.append(delta)
    
df5_delta_data.pop(0)

np.mean(df5_delta_data)

0.7094999545454548

### Our KLM operator results

For using keyboard only interaction techniques we conclude with:

- pressing a single key uniquely : 0.25342900000000007
- pressing a shift single key: 0.545387125
- pressing a single key multiple times: 0.2338534081632653

For using mouse only interaction techniques we conclude with:

- clicking a single digit after a digit uniquely: 0.5449099655172415
- clicking a single arithmetic operation after a digit: 0.7094999545454548